include/deal.II/differentiation/ad/ad_helpers.h

```diff
@@ -709,13 +731,344 @@ namespace Differentiation
+     * deal.II, so this class should @b not be embedded within a multithreaded
+     * function when using ADOL-C number types. It is, however, suitable for use
+     * in both serial and MPI routines.
+     *
+     * @author Jean-Paul Pelteret, 2016, 2017, 2018
+     */
+    template <enum AD::NumberTypes ADNumberTypeCode,
+              typename ScalarType = double>
+    class ADHelperCellLevelBase
+      : public ADHelperBase<ADNumberTypeCode, ScalarType>
+    {
+    public:
+      /**
+       * Type definition for the floating point number type that are used in,
+       * and result from, all computations.
+       */
+      using scalar_type =
+        typename ADHelperBase<ADNumberTypeCode, ScalarType>::scalar_type;
+
+      /**
+       * Type definition for the auto-differentiation number type that is used
+       * in all computations.
+       */
+      using ad_type =
+        typename ADHelperBase<ADNumberTypeCode, ScalarType>::ad_type;
+
+      /**
+       * @name Constructor / destructor
+       */
+      //@{
+
+      /**
+       * The constructor for the class.
+       *
+       * @param[in] n_independent_variables The number of independent variables
+       * that will be used in the definition of the functions that it is
+       * desired to compute the sensitivities of. In the computation of
+       * $\mathbf{f}(\mathbf{X})$, this will be the number of inputs
+       * $\mathbf{X}$, i.e. the dimension of the range space.
+       * @param[in] n_dependent_variables The number of scalar functions to be
+       * defined that will have a sensitivity to the given independent
+       * variables. In the computation of $\mathbf{f}(\mathbf{X})$, this will
+       * be the number of outputs $\mathbf{f}$, i.e. the dimension of the
+       * domain or image space.
+       */
+      ADHelperCellLevelBase(const unsigned int n_independent_variables,
+                            const unsigned int n_dependent_variables);
+
+      /**
+       * Destructor
+       */
+      virtual ~ADHelperCellLevelBase();
+
+      //@}
+
+      /**
+       * @name Independent variables
+       */
+      //@{
+
+      /**
+       * Register the complete set of independent variables $\mathbf{X}$ that
+       * represent the local degree-of-freedom values.
+       *
+       * @param[in] dof_values A vector field associated with local
+       * degree-of-freedom values on the current finite element. These define
+       * the values of all independent variables. When considering taped AD
+       * numbers with branching functions, to avoid potential issues with branch
+       * switching it may be a good idea to choose these values close to those
+       * that will be later evaluated and linearized around.
+       *
+       * @note The input value type must correspond to this class's number type.
+       * So if this class is templated on type double then the number type
+       * associated with value must also be of type double.
```

Comment:
Hm, this class *is* templated. Do you mean to say "if the ... template argument of this class is @p double, then ..."?

Also, that's not how `scalar_type` is declared. It is not equal to the template argument of this class, but something more complicated.

Reply:
Indeed. I've now reformulated the message that I'm trying to convey here.